tutorials/application_design/py/subscriber.py

```diff
@@ -45,71 +38,110 @@ def __init__(
     ):
         self._metrics_reader = metrics_reader
         self._transit_reader = transit_reader
-        self._dashboard_data: typing.Dict[
-            dds.InstanceHandle, DashboardItem
-        ] = dict()
 
     def __repr__(self):
-        return f"Dashboard({self._metrics_reader=}, {self._transit_reader=}, {self._dashboard_data=})"
+        return f"Dashboard({self._metrics_reader=}, {self._transit_reader=})"
+
+    def run(self):
+
+        def new_position_handler(_):
+            print(self._new_position_string(new_position_condition))
 
-    async def run(self):
-        await asyncio.gather(
-            self._display_app(),
-            self._metrics_app(),
-            self._transit_app(),
+        new_position_condition = dds.ReadCondition(
+            self._transit_reader, dds.DataState.new_data, new_position_handler
         )
 
-    @property
-    def online_vehicles(self):
-        return {
-            handle: data
-            for handle, data in self._dashboard_data.items()
-            if not data.is_historical
-        }
-
-    @property
-    def offline_vehicles(self):
-        return {
-            handle: data
-            for handle, data in self._dashboard_data.items()
-            if data.is_historical
-        }
-
-    async def _display_app(self):
-        while True:
-            print(f"[[ DASHBOARD: {datetime.now()} ]]")
-            print(f"Online vehicles: {len(self.online_vehicles)}")
-            for data in self.online_vehicles.values():
-                print(f"- Vehicle {data.vin}:")
-                print(f"  Fuel updates: {len(data.fuel_history)}")
-                print(f"  Last known destination: {data.current_destination}")
-                print(f"  Last known fuel level: {data.fuel_history[-1]}")
-            print(f"Offline vehicles: {len(self.offline_vehicles.keys())}")
-            for data in self.offline_vehicles.values():
-                mean_full_consumption = statistics.mean(
-                    data.fuel_history
-                ) / len(data.fuel_history)
-
-                print(f"- Vehicle {data.vin}:")
-                print(f"  Mean fuel consumption: {mean_full_consumption}")
-                print(
-                    f"  Known reached destinations: {len(data.reached_destinations)}"
+        dashboard_condition = dds.GuardCondition()
+
+        def dashboard_handler(_):
+            print(self._dashboard_string())
+
+        dashboard_condition.set_handler(dashboard_handler)
+
+        display_handler_sentinel = threading.Event()
+        display_handler_condition = threading.Condition()
+
+        def display_handler():
+            while not display_handler_sentinel.is_set():
+                with display_handler_condition:
+                    display_handler_condition.wait(5)
+                    dashboard_condition.trigger_value = True
```

Comment:
can you explain why you use a GuardCondition instead of just calling `print(self._dashboard_string())` here?

Reply:
If I did not use a Guard Condition and instead printed after every dispatch, I would always get the dashboard printed whenever the waitset is dispatched.

Using a guard condition, I can make it so that the Guard Condition is printed consistely over time.